source/io/module_javaScriptInvoke.js

```diff
@@ -74,32 +74,121 @@ var assignJavaScriptInvoke;
-            return jsValueToCookieCache.get(jsValue);
-        };
+            get (cookie) {
+                var jsValue = this._cookieToJsValueMap.get(cookie);
+                if (jsValue === undefined) {
+                    throw new Error('Attempted to get a Static JavaScript Refnum with a cookie that has not been set.');
+                }
+                return jsValue;
+            }
 
-        var generateUniqueRefNumCookie = function () {
-            jsRefNumCookieCounter += 1;
-            return jsRefNumCookieCounter;
-        };
+            has (cookie) {
+                return this._cookieToJsValueMap.has(cookie);
+            }
+
+            get size () {
+                return this._cookieToJsValueMap.size;
+            }
+        }
+
+        class DynamicRefnumManager {
+            constructor () {
+                this._cookieToJsValueMap = new Map();
+            }
+
+            set (jsValue) {
+                // Any value allowed and always creates a new cookie
+                var cookie = generateUniqueRefNumCookie();
+                this._cookieToJsValueMap.set(cookie, jsValue);
+                return cookie;
+            }
+
+            setMultiple (values, cookies) {
+                var i, cookie;
+                for (i = 0; i < values.length; i += 1) {
+                    cookie = generateUniqueRefNumCookie();
+                    this._cookieToJsValueMap.set(cookie, values[i]);
+                    cookies[i] = cookie;
+                }
+            }
+
+            get (cookie) {
+                // Have to check the map because undefined and null are possible values for dynamic references
+                if (!this._cookieToJsValueMap.has(cookie)) {
+                    throw new Error('Attempted to get a Dynamic JavaScript Refnum with a cookie that has not been set.');
+                }
+                var jsValue = this._cookieToJsValueMap.get(cookie);
+                return jsValue;
+            }
+
+            getMultiple (cookies, values) {
+                var i;
+                for (i = 0; i < cookies.length; i += 1) {
+                    if (this._cookieToJsValueMap.has(cookies[i])) {
+                        values[i] = this._cookieToJsValueMap.get(cookies[i]);
+                    } else {
+                        return false;
+                    }
+                }
+                return true;
+            }
+
+            has (cookie) {
+                return this._cookieToJsValueMap.has(cookie);
+            }
+
+            delete (cookie) {
+                return this._cookieToJsValueMap.delete(cookie);
+            }
+
+            get size () {
+                return this._cookieToJsValueMap.size;
+            }
+        }
+
+        var staticRefnumManager = new StaticRefnumManager();
+        var dynamicRefnumManager = new DynamicRefnumManager();
+
+        Module.javaScriptInvoke.do_not_use_debug_only_static_refnum_manager = staticRefnumManager;
```

Comment:
Placed on the internal module to make available in tests but should not be used by production code